### PR TITLE
Make Oid zero copy

### PR DESCRIPTION
**Pull request by jannschu**
Hi,

I implemented the changes discussed in #17.

See the README for a changelog.

I added a crate for the procedural macro. The Cargo.toml needs some metadata. I am not sure if cargo can deploy it like this?

What do you think about the changes?

### CODE REVIEW

**Comment by nhynes** [comment]
@jannschu 👋

**Comment by chifflier** [comment]
Hi, sorry for the delay. I'm catching up the proposed pull requests
Unfortunately, cargo has no support for subcrates (nor publishing in workspaces), which means it would have to be published as an additional crate (even if in the same source), or distributed inside the cargo package (which may be easier).

This is quite a huge refactor of OID handling, and while some changes looks really interesting I will need some time to have a look

File: src/oid.rs [hunk 1]
```diff
@@ -1,62 +1,228 @@
...
+        pub fn to_string(&self) -> String {
+            let ints: Vec<String> = self.iter().map(|i| i.to_string()).collect();
+            ints.join(".")
+        }
 
-        for it in self.0.iter().skip(1) {
-            s.push('.');
-            s = s + &it.to_string();
+        /// Return an iterator over the sub-identifiers (arcs).
+        #[cfg(feature = "bigint")]
```
Comment (chifflier) [inline src/oid.rs]: This is redundant (module is already feature-guarded)

File: der-oid-macro/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+		multi::{many0_count, separated_list},
+		sequence::{delimited, pair, terminated, tuple},
+		IResult, exact, call,
+	};
+	
+	fn uint<'a, E: ParseError<&'a str>>(i: &'a str) -> IResult<&'a str, &'a str, E> {
+		recognize::<&str, &str, E, _>(
+			alt((
+				map(char('0'), |_| ""),
+				map(pair(one_of("123456789"), many0_count(one_of("0123456789"))), |_| "")
```
[the remark this answers is not in the document]
Comment (chifflier) [reply der-oid-macro/src/lib.rs]: Also, won't that fail if number is written with underscores ? (for ex `123_456`)

**Comment by jannschu** [comment]
Hi,

thanks for taking the time to review the changes. I was on vacation.

> Overall I'm not against this patch, but if I'm wondering what the real gain is, or what problem was.
> Using the raw bytes trades parsing time against access time, and prevents accessing directly at some index to get a component, forcing to iterate every time.
> I'm not entirely sure if that would help, for ex in X.509

I think the representation like 1.2.234.23 is rarely needed. Even advanced users can not remember those. It might be useful for debugging, but even developers might prefer the name instead of the number, depending on what software you are working on. 

So my argument is that the most common operation done with OIDs is checking if they are equal to  another oid. For this purpose it is wasteful to decode it. It is sufficient to compare the encoded form, given that they are unique, which is the case for DER. Whether uniqueness is guaranteed for valid BER is something which needs to be checked, technically it is possible to add leading zeros.

Regarding the bigint dependency: A u64 iterator should fulfill all relevant use cases, I guess. The current implementation on master can be crashed purposefully by giving it an oid which is too long. This needs to be addressed in my opinion.  A check as you suggest should work. I will provide an optional bigint iterator and a u64 with `Result<Iterator, TooBig>` return type.

**Comment by jannschu** [comment]
I addressed the review comments and rebased on master.

**Comment by jannschu** [comment]
> Whether uniqueness is guaranteed for valid BER is something which needs to be checked, technically it is possible to add leading zeros.

I just searched in X.690. In 8.20/8.21 the document explicitly requires that no leading zeros are allowed. So it is safe to assume this for an implementation in my opinion.

**Comment by chifflier** [comment]
Thanks, I'll have a look ASAP

**Comment by jannschu** [comment]
I rebased on master, used rustfmt, fixed clippy warnings and squashed everything.

**Comment by jannschu** [comment]
To the best of my knowledge internal path dependencies are not supported at the moment (affects oid macro). The typical solution seems to publish the implementation crate. This means, `der-oid-macro` would need to be become a public crate.

**Comment by chifflier** [comment]
Hi,
I'd like to release a new version of the crate, with this PR if possible. I'm testing this PR to see if the API changes and the embedded crate are easy to use.

**Comment by chifflier** [comment]
I pulled the PR on a local branch and resolved the conflicts.
I am currently testing the impact of the changes on other crates (like `x509-parser`)

**Comment by chifflier** [comment]
Ok, so I decided to merge (resolving conflicts and running rustfmt) in 6097b31
The upgrade path is usually not difficult (add a few `unwrap()` and lifetimes, it was just a bit more painful in `x509-parser` due to the `Oid` registry being `'static'. It required some refactoring, but nothing too difficult.

Thanks for your PR!

**Pull request merged**